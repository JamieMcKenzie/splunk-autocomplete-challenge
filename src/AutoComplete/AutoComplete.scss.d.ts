interface CssExports {
  'wrapper': string;
  'control': string;
  'isLoading': string;
}

export const cssExports: CssExports;
export default cssExports;
