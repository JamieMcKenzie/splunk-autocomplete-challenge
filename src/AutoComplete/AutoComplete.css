.wrapper {
  position: relative;
  max-width: 50%;
  margin: 20px auto;
}

.input-field {
  width: 100%;
}

.list {
  width: 100%;
  display: flex;
  flex-direction: column;
  background-color: rgb(238, 234, 234);
}

a.list-item {
  padding: 10px;
  color: rgb(67, 102, 148);
  text-align: left;
  border-bottom: .1px solid lightgray;
  cursor: pointer;
}

.input-container {
  height: 32px;
  width: 100%;
}

.input-field {
  height: 32px;
  width: 100%;
}

.icon-container {
  display: none;
  position: absolute;
  right: 2px;
  top: 5px;
}

.is-loading .icon-container {
  display: block;
}

.loader {
  position: relative;
  height: 20px;
  width: 20px;
  display: inline-block;
  animation: around 10s infinite;
}

/* Exercise specified that CSS was not a part of the evaluation, so I grabbed this wheel animation from an online CSS example and tweaked it a little */
@keyframes around {
  0% {
    transform: rotate(0deg)
  }
  100% {
    transform: rotate(360deg)
  }
}

.loader::after, .loader::before {
  content: "";
  background: white;
  position: absolute;
  display: inline-block;
  width: 100%;
  height: 100%;
  border-width: 2px;
  border-color: #333 #333 transparent transparent;
  border-style: solid;
  border-radius: 20px;
  box-sizing: border-box;
  top: 0;
  left: 0;
  animation: around 0.5s ease-in-out infinite;
}

.loader::after {
  animation: around 0.5s ease-in-out 0.1s infinite;
  background: transparent;
}