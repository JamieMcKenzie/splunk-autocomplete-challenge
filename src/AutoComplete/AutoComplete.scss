.wrapper {
  position: relative;
}

.control {
  background-color: red;
}

.isLoading {
  background-color: blue;
}
