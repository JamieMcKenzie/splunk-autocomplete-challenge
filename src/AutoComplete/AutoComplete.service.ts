import { AxiosInstance } from 'axios';
import _ from 'lodash';
const axios = require('axios').default;

const axiosInstance = axios.create({
  baseURL: 'https://dry-cliffs-99602.herokuapp.com/api',
  timeout: 1000,
  headers: {
    'Content-Type': 'application/x-www-form-urlencoded',
    'Access-Control-Allow-Origin': '*',
  }
});

const getAxiosInstance = (): AxiosInstance => {
  return axiosInstance
}

export const getSearchSuggestions = (userQuery: string): Promise<string[]> => {
  let results: string[];
  return axiosInstance({
    method: 'get',
    url: '/items',
    params: {
      q: userQuery,
    }
  }).
  then((response: any) => {
    debugger;
    return response.data;
  }).catch((e: object) => {
    return ['error'];
  });
};

