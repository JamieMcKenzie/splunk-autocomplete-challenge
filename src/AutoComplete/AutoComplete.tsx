import cx from 'classnames';

import styles from './AutoComplete.scss';

interface IAutoCompleteProps {
  suggestions: string[];
  isLoading: boolean;
}

const AutoComplete: React.FC<IAutoCompleteProps> = ({ suggestions, isLoading }) => {
  const controlClassnames = cx(styles.control, {
    [styles.isLoading]: isLoading,
  });


  return (
    <div className='wrapper'>
      <div className='controlClassnames'>
        <input />
      </div>
      <div className='list'>
        {suggestions && suggestions.map((suggestion: string, index: number): JSX.Element => {
          return (
            <li key={index} className='list-item'>
              {suggestion}
            </li>
          );
        })}
      </div>
    </div>
  );

};

export default AutoComplete;
