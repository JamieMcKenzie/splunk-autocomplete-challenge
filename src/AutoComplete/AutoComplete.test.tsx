import { cleanup } from '@testing-library/react'
import { mount, shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

import * as AutoCompleteService from './AutoComplete.service'
import AutoComplete from './AutoComplete'
import { doesNotReject } from 'assert'

afterEach(cleanup)
const mockSelectFunc = jest.fn()
jest.mock('./AutoComplete.service', () => {
    return {
        getSearchSuggestions: jest.fn()
    }
})
const mockedData = [ 'result1', 'result2', 'result3' ]
const mockResponse = { data: mockedData }
const event = { target: { value: 'query' } } as React.ChangeEvent<HTMLInputElement>

describe('snapshot of AutoComplete', () => {
    it("renders correctly", () => {
        const tree = shallow(<AutoComplete onSelectItem={ mockSelectFunc }/>)
        expect(toJSON(tree)).toMatchSnapshot()
    })
})

describe('AutoComplete Component', () => {
    const wrapper = mount(<AutoComplete onSelectItem={ mockSelectFunc } />)
    
    it('renders with onSelectItem prop', () => {
        const props = { onSelectItem: mockSelectFunc }
        expect(wrapper.props().onSelectItem).toBeDefined
    })

    it('state contains isLoading and suggestions', () => {
        expect(wrapper.state('isLoading')).toBe(false)
        expect(wrapper.state('suggestions')).toBeUndefined
    })

    it('should not render list if no suggestions', () => {
        expect(wrapper.find('.list')).toHaveLength(0)
    })

    it('should render list is suggestions are populated', () => {
        wrapper.setState({ suggestions: mockedData })
        expect(wrapper.find('.list')).toHaveLength(1)
        expect(wrapper.find('.list-item')).toHaveLength(3)
    })

    it('should render input field', () => {
        expect(wrapper.find('input')).toHaveLength(1)
    })
})

describe('onSelectItem Func', () => {
    const wrapper = mount(<AutoComplete onSelectItem={ mockSelectFunc } />)
    wrapper.setState({
        isLoading: false,
        suggestions: mockedData,
    })
    it('should render selectable list items if suggestions is populated', () => {
        const suggestion = wrapper.find('.list-item')
        expect(suggestion).toHaveLength(3)
        suggestion.at(0).simulate('click', event )
        expect(mockSelectFunc.mock.calls.length).toEqual(1);
    })
})

describe('AutoCompleteService - getSearchSuggestions', () => {
    (AutoCompleteService.getSearchSuggestions as jest.MockedFunction<typeof AutoCompleteService.getSearchSuggestions>).mockResolvedValueOnce(
        mockedData
      )
      const wrapper = shallow(<AutoComplete onSelectItem={ mockSelectFunc } />)
      setTimeout(() => {
        expect(wrapper.state('suggestions')).toEqual(mockedData)
        expect(wrapper).toMatchSnapshot();
      }, 0);
})
