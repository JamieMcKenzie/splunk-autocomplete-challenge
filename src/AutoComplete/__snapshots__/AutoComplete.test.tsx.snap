// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of AutoComplete renders DOM correctly 1`] = `
<div
  className="wrapper"
>
  <div
    className=" control"
  >
    <div
      className="input-container"
    >
      <input
        autoComplete="off"
        className="input-field"
        id="search-input"
        name="search"
        onChange={[Function]}
        placeholder="Search"
        type="search"
      />
      <div
        className="icon-container"
      >
        <i
          className="loader"
        />
      </div>
    </div>
  </div>
</div>
`;
